### PR TITLE
GetTransactionDetails: Properly extract Options from PaymentItemInfo; also, document and test Options clearly

### PR DESCRIPTION
**Pull request by bkuhn**
*This pull request is an updated version of Pull Request #4.  Pull Request #4 was automatically closed by GitHub because I rebased the branch.  (Why can't GitHub support that?)  Anyway, I also added a few new things as well.*

I recently attempted to make use of the Options fields as part of PaymentItemInfo and discovered that the API didn't provide the actual data regarding the Options via its interface.

This pull request corrects that bug in the API. It does make a minor API-user-visible change that I think is appropriate given the nature of the bug. I've fully documented that in the Changes file and in the API documentation.

Finally, I hereby license my copyrights herein under the same license as Perl itself. :)

### CODE REVIEW

**Comment by oalders** [comment]
Thanks.  I'll check this out at some point over the weekend.  How would you feel about getting co-maint on this module and releasing the next version to CPAN?  I don't want to be a blocker on getting these changes out.

File: lib/Business/PayPal/API/GetTransactionDetails.pm [hunk 1]
```diff
@@ -301,3 +349,13 @@ at your option, any later version of Perl 5 you may have available.
...
+
+# Local Variables:
+#   Mode: CPerl
+#   indent-tabs-mode: nil
+#   cperl-indent-level: 4
+#   cperl-brace-offset: 0
+#   cperl-continued-brace-offset: 0
+#   cperl-label-offset: -4
+#   cperl-continued-statement-offset: 4
+# End:
```
Comment (oalders) [inline lib/Business/PayPal/API/GetTransactionDetails.pm]: From what I gather these are hints to emacs?  I don't think they need to be in this file.  I'd rather see a .perltidyrc file included with the distribution so that all of the formatting requirements can be in one place.  I can add one once this pull request has been merged.

Comment (oalders) [inline lib/Business/PayPal/API/GetTransactionDetails.pm]: See also https://github.com/autarch-code/perl-code-tidyall/blob/master/etc/editors/tidyall.el

File: t/OptionFields.t [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+# Local Variables:
+#   Mode: CPerl
+#   indent-tabs-mode: nil
+#   cperl-indent-level: 4
+#   cperl-brace-offset: 0
+#   cperl-continued-brace-offset: 0
+#   cperl-label-offset: -4
+#   cperl-continued-statement-offset: 4
+# End:
```
Comment (oalders) [inline t/OptionFields.t]: See above comment regarding cperl config.

File: t/OptionFields.t [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+and use the sandbox buyer account to make the payment.
+_OPTIONS_LINK_
+}
+
+
+my $startdate = '1998-01-01T01:45:10.00Z';
+
+my $ts   = new Business::PayPal::API::TransactionSearch( %args );
+my $td   = new Business::PayPal::API::GetTransactionDetails( %args );
```
Comment (oalders) [inline t/OptionFields.t]: `Business::PayPal::API::GetTransactionDetails->new( %args )`

File: t/OptionFields.t [hunk 4]
```diff
@@ -0,0 +1,133 @@
+# This file is part of Business:PayPal:API Module.   License: Same as Perl.  See its README for details.
+# -*- mode: cperl -*-
+use Test::More;
+use strict;
```
Comment (oalders) [inline t/OptionFields.t]: We should `use warnings` as well.

File: t/OptionFields.t [hunk 5]
```diff
@@ -0,0 +1,133 @@
+# This file is part of Business:PayPal:API Module.   License: Same as Perl.  See its README for details.
+# -*- mode: cperl -*-
+use Test::More;
```
Comment (oalders) [inline t/OptionFields.t]: The `strict` and `warnings` pragmas don't kick in until after they are used, so they should precede any other use statements.

File: t/OptionFields.t [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+if ( !$ENV{WPP_TEST} || !-f $ENV{WPP_TEST} ) {
+    plan skip_all =>
+        'No WPP_TEST env var set. Please see README to run tests';
+}
+else {
+    plan tests => 14;
+}
+
+use_ok( 'Business::PayPal::API::TransactionSearch' );
```
Comment (oalders) [inline t/OptionFields.t]: I think `use_ok` is probably not a useful test here.  See http://modernperlbooks.com/mt/2012/04/use-ok-is-broken-because-require-is-broken.html

File: t/OptionFields.t [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+my $ts   = new Business::PayPal::API::TransactionSearch( %args );
+my $td   = new Business::PayPal::API::GetTransactionDetails( %args );
+
+my $resp = $ts->TransactionSearch(StartDate     => $startdate);
+my %detail;
+
+foreach my $record (@{$resp}) {
+    %detail = $td->GetTransactionDetails(TransactionID => $record->{TransactionID});
+    last if $detail{PII_Name} =~ /$itemName/;
+    %detail = {};
```
Comment (oalders) [inline t/OptionFields.t]: What is the purpose of this line?

File: t/OptionFields.t [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+expecting only the first two back.
+
+More details on this can be found in the
+L<Business:PayPal:API:GetTransactionDetails/"PaymentItem Options Limitations">
+documentation.
+
+=cut
+
+foreach my $options ($detail{PaymentItems}[0]{Options}, $detail{PII_Options}[0]) {
+    ok(scalar(keys %$options) == 2, "The PaymentItems Options has 2 elements");
```
Comment (oalders) [inline t/OptionFields.t]: `is( scalar keys %{$options}, 2, 'The PaymentItems Options has 2 elements' )`

File: t/OptionFields.t [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    ok(defined $options->{firstOption}, "'firstOption' is present");
+    ok($options->{firstOption} eq 'Yes', "'firstOption' is selected as 'Yes'");
```
Comment (oalders) [inline t/OptionFields.t]: I would use `is` here as well.

File: t/OptionFields.t [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    ok(defined $options->{size}, "'size' option is present");
+    ok($options->{size} eq "Large", "'size' option is selected as 'Large'");
```
Comment (oalders) [inline t/OptionFields.t]: Also `is`

File: t/OptionFields.t [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+=pod
+
+These tests verify the options work.
+
+=cut
+
+my $itemName;
+
+if (-f 'options-payment.html') {
+    open(OPTIONS_PAY_HTML_READ, "<", "options-payment.html");
```
Comment (oalders) [inline t/OptionFields.t]: No need to use brackets when the function is a built-in.  Don't change this if you don't feel like it, but stylistically I think it's fine to omit them.

**Comment by oalders** [comment]
Thanks for all of your work on this.  I did leave a bunch of comments.  They're little things, but I think they're worth addressing.  The problem is that the rest of the code is in crufty state, so it's probably a good idea to make sure the new additions don't follow the same patterns where possible.

**Comment by bkuhn** [comment]
@oalders All your patch review looks really excellent.  I'm sorry that my Perl is a bit rusty: it's been basically 15 years since I was doing Perl every day, so various parts are a bit rusty, and I admit to a certain amount of embarrassment upon reading your patch review.  I can definitely fix everything you suggest and resubmit, and you should feel free to wait to merge until I can do it, although it may be about two weeks before I have a chance, b/c I'm headed to LinuxConf AU next week.  (Probably seeing Nat Torkington for the first time in years, so it will help spark my Perl memories. ;)

As for co-maintainer, my fear is that I won't have adequate, and I'll just be unreliable.  While I do rely on this module every day, OTOH it mostly does what I need.   What I'm pushing you now is basically my last 3 years of work which I never properly upstreamed.  But it's a spurt of stuff, not a new regular flow of work.  Let me think on it and maybe we should discuss on IRC later in the month to specifically discuss what you need me to take on.

File: t/OptionFields.t (hunk 7 above)
Comment (bkuhn) [inline t/OptionFields.t]: This is to make sure %detail is empty at the end of the loop if we *don't* find the specific one that matches $itemName.  Can you suggest a better way to do this?

**Comment by oalders** [comment]
@bkuhn let me just say that it's not _on any level_ my intention for you to feel badly about your code.  I think it's great that you're putting so much time into this and I'm quite happy about the fact that you're providing appropriate tests.  I got lots of pull requests without tests, so I already feel spoiled with your code. I just thought I'd pass along some helpful hints as many others have done for me in the past.  I was wondering if maybe I hammered you on this last pull request, but it was really just little stuff and me being opinionated.  Nothing to feel badly about.  ;)

As far as the next release goes, I'm just a bit tight for time.  If you were able to confirm that the sandbox tests are now no worse off than they were at the previous release then I'd be happy to cut a new release and see what the fallout is.  I just thought that since you're showing some interest, that I wouldn't want to stand in your way if you wanted to take this on yourself.

So, basically, if you can give me some feedback on the sandbox tests, we can get this moving, but if you do feel like this is something you want more control of, I certainly won't stand in your way.

Thanks again for this and have a great time at LinuxConf AU.

File: t/OptionFields.t (hunk 7 above)
Comment (oalders) [inline t/OptionFields.t]: Ah, in that case I think you meant `%detail = ()`.

I _think_ what you are trying to do would be clearer if you changed the `last if` in line 91 to a `next unless` and then moved the `like` tests inside of the `foreach` loop.  That way you could contain the scope of `%detail` inside the `foreach` loop and there should be no ambiguity about what you're doing.  Does that make sense?

**Comment by oalders** [comment]
@bkuhn thanks so much for this.  I've merged it as-is.  I think we're both pretty busy and I did want to get your changes in!  I'm going to close this pull request as there were merge conflicts, so I handled this manually.
